sql/2020/11_Security/security_headers_prevalence.sql

```diff
@@ -0,0 +1,34 @@
+#standardSQL
+# Prevalence of security headers set in a first-party context; count by number of hosts.
+CREATE TEMPORARY FUNCTION hasHeader(headers STRING, headername STRING)
+  RETURNS BOOL DETERMINISTIC
+  LANGUAGE js AS '''
+  const parsed_headers = JSON.parse(headers);
+  const matching_headers = parsed_headers.filter(h => h.name.toLowerCase() == headername.toLowerCase());
+  return matching_headers.length > 0;
+''';
+
+SELECT
+  client,
+  headername,
+  COUNT(DISTINCT host) AS total_hosts,
+  COUNT(DISTINCT IF(hasHeader(headers, headername), host, NULL)) AS num_with_header,
+  COUNT(DISTINCT IF(hasHeader(headers, headername), host, NULL)) / COUNT(DISTINCT host) AS pct_with_header
+FROM (
+  SELECT
+    client,
+    NET.HOST(urlShort) AS host,
+    JSON_EXTRACT(payload, '$.response.headers') AS headers
+  FROM
+    `httparchive.almanac.requests`
+  WHERE
+    date = "2020-08-01" AND
+    NET.HOST(urlShort) = NET.HOST(page)
+),
+UNNEST(['Content-Security-Policy', 'Cross-Origin-Embedder-Policy', 'Cross-Origin-Opener-Policy', 'Cross-Origin-Resource-Policy', 'Feature-Policy', 'Referrer-Policy', 'Report-To', 'Strict-Transport-Security', 'X-Content-Type-Options', 'X-Frame-Options', 'X-XSS-Protection']) AS headername
```

Comment:
Should we measure the new [Permissions Policy](https://scotthelme.co.uk/goodbye-feature-policy-and-hello-permissions-policy/)? Not supported by browsers yet but with securityheaders.com pushing it, might be interesting to see.

```suggestion
UNNEST(['Content-Security-Policy', 'Cross-Origin-Embedder-Policy', 'Cross-Origin-Opener-Policy', 'Cross-Origin-Resource-Policy', 'Feature-Policy', 'Permissions-Policy', 'Referrer-Policy', 'Report-To', 'Strict-Transport-Security', 'X-Content-Type-Options', 'X-Frame-Options', 'X-XSS-Protection']) AS headername
```